From c7097d9b185144956ea864a46fd67839d746f6ca Mon Sep 17 00:00:00 2001
From: Yunguo Wei <yunguo.wei@windriver.com>
Date: Thu, 18 Jan 2018 16:42:26 +0800
Subject: [PATCH 4/5] rsi: use kmalloc mem for usb msg buffer

This patch is addressing the following issue:

[    3.498933] transfer buffer not dma capable
[    3.498953] ------------[ cut here ]------------
[    3.498963] WARNING: CPU: 0 PID: 276 at
/kernel-source//drivers/usb/core/hcd.c:1588 usb_hcd_map_urb_for_dma+0x37e/0x570
[    3.498964] Modules linked in: rsi_usb(+) intel_rapl x86_pkg_temp_thermal intel_powerclamp coretemp rsi_91x kvm_intel
rsi_zigb kvm irqbypass mac80211 efiv ars(+) i2c_designware_platform cfg80211 spi_pxa2xx_platform i2c_designware_core igb(+) i915 snd_soc_skl snd_soc_skl_ipc
snd_soc_sst_ipc snd_soc_sst_dsp snd_h da_ext_core snd_soc_sst_match mei_me snd_hda_intel mei snd_hda_codec
snd_hda_core snd_soc_rt298 snd_soc_rt286 snd_soc_rl6347a snd_soc_core snd_compress ac97_bus snd_pcm snd_timer video ip_tables x_tables
[    3.498998] CPU: 0 PID: 276 Comm: systemd-udevd Not tainted 4.12.18-pulsar-preempt-rt #1
[    3.498999] Hardware name: ADLINK TECHNOLOGY Inc. MXE210/MXE210, BIOS 0.14.10 10/31/2017
[    3.499000] task: ffff99a53630c300 task.stack: ffff9afa40754000
[    3.499003] RIP: 0010:usb_hcd_map_urb_for_dma+0x37e/0x570
[    3.499005] RSP: 0018:ffff9afa407578a8 EFLAGS: 00010296
[    3.499007] RAX: 000000000000001f RBX: ffff99a5741eb780 RCX:0000000000000001
[    3.499008] RDX: 0000000080000001 RSI: 0000000000000002 RDI:00000000ffffffff
[    3.499009] RBP: ffff9afa407578e8 R08: 0000000000000000 R09:000000000000001f
[    3.499010] R10: ffff99a5741eb780 R11: 000000000000028b R12:00000000fffffff5
[    3.499011] R13: 0000000001400000 R14: 0000000000000002 R15:ffff99a576a68000
[    3.499013] FS:  00007f796a889240(0000) GS:ffff99a572400000(0000) knlGS:0000000000000000
[    3.499014] CS:  0010 DS: 0000 ES: 0000 CR0: 0000000080050033
[    3.499015] CR2: 00007f796a889158 CR3: 0000000075e47000 CR4:00000000003406f0
[    3.499016] Call Trace:
[    3.499022]  usb_hcd_submit_urb+0x3e5/0x9c0
[    3.499026]  ? _raw_spin_lock+0x17/0x40
[    3.499031]  ? sched_clock_cpu+0x1a/0xe0
[    3.499033]  ? preempt_count_add+0x74/0xc0
[    3.499036]  usb_submit_urb+0x33b/0x510
[    3.499039]  ? debug_smp_processor_id+0x17/0x20
[    3.499041]  usb_start_wait_urb+0x5f/0xe0
[    3.499044]  usb_control_msg+0xc1/0x100
[    3.499051]  rsi_usb_reg_read.constprop.12+0x43/0x80 [rsi_usb]
[    3.499055]  rsi_probe+0x56f/0x770 [rsi_usb]
[    3.499059]  usb_probe_interface+0xee/0x2a0
[    3.499062]  driver_probe_device+0x226/0x2e0
[    3.499064]  __driver_attach+0x91/0xa0
[    3.499066]  ? driver_probe_device+0x2e0/0x2e0

Signed-off-by: Yunguo Wei <yunguo.wei@windriver.com>
---
 drivers/net/wireless/rsi/rsi_91x_usb.c | 14 ++++++++++++--
 1 file changed, 12 insertions(+), 2 deletions(-)

diff --git a/drivers/net/wireless/rsi/rsi_91x_usb.c b/drivers/net/wireless/rsi/rsi_91x_usb.c
index 27bf76d..77fe46c 100644
--- a/drivers/net/wireless/rsi/rsi_91x_usb.c
+++ b/drivers/net/wireless/rsi/rsi_91x_usb.c
@@ -162,11 +162,15 @@ static int rsi_usb_reg_read(struct usb_device *usbdev,
 			    u32 *value,
 			    u16 len)
 {
-	u8 buf[4];
+	u8 *buf;
 	int status = 0;
 	u16 reg_value;
 	u16 index;
 
+	buf = kmalloc(4, GFP_KERNEL);
+	if (!buf)
+		return -ENOMEM;
+
 	len = 2;
 	reg_value = cpu_to_le16(((u16 *)&reg)[1] & 0xffff);
 	index = cpu_to_le16(((u16 *)&reg)[0] & 0xffff);
@@ -187,6 +191,7 @@ static int rsi_usb_reg_read(struct usb_device *usbdev,
 			__func__, status);
 	}
 
+	kfree(buf);
 	return status;
 }
 
@@ -205,10 +210,14 @@ static int rsi_usb_reg_write(struct usb_device *usbdev,
 			     unsigned long value,
 			     u16 len)
 {
-	u8 usb_reg_buf[4];
+	u8 *usb_reg_buf;
 	int status = 0;
 	u16 reg_value, index;
 
+	usb_reg_buf = kmalloc(4, GFP_KERNEL);
+	if (!usb_reg_buf)
+		return -ENOMEM;
+
 	usb_reg_buf[0] = (value & 0x00ff);
 	usb_reg_buf[1] = (value & 0xff00) >> 8;
 	usb_reg_buf[2] = 0x0;
@@ -231,6 +240,7 @@ static int rsi_usb_reg_write(struct usb_device *usbdev,
 			__func__, status);
 	}
 
+	kfree(usb_reg_buf);
 	return status;
 }
 
-- 
2.7.4

